### PR TITLE
Client channel init cleanup

### PR DESCRIPTION
**Pull request by markdroth**
Instead of requiring an out-of-band call to `grpc_client_channel_finish_initialization()`, we pass in the server URI and client channel factory via channel args, and we instantiate the resolver directly in client_channel's `init_channel_elem()` method.

Note that this also changes the API of `init_channel_elem()` to return a `grpc_error*`, so that we can indicate when we fail to instantiate the resolver.

### CODE REVIEW

File: src/core/lib/channel/channel_stack_builder.c [hunk 1]
```diff
@@ -250,15 +249,15 @@ void *grpc_channel_stack_builder_finish(grpc_exec_ctx *exec_ctx,
   size_t channel_stack_size = grpc_channel_stack_size(filters, num_filters);
 
   // allocate memory, with prefix_bytes followed by channel_stack_size
-  char *result = gpr_malloc(prefix_bytes + channel_stack_size);
+  *result = gpr_malloc(prefix_bytes + channel_stack_size);
   // fetch a pointer to the channel stack
   grpc_channel_stack *channel_stack =
-      (grpc_channel_stack *)(result + prefix_bytes);
+      (grpc_channel_stack *)(*result + prefix_bytes);
```
Comment (y-zeng) [inline src/core/lib/channel/channel_stack_builder.c]: We may need `(grpc_channel_stack *)((char *)(*result) + prefix_bytes)` here to make windows happy.

Comment (markdroth) [inline src/core/lib/channel/channel_stack_builder.c]: Done.

**Comment by ctiller** [comment]
Jenkins: test this please

**Comment by ctiller** [comment]
Seems to break grpclb test on asan

File: include/grpc/impl/codegen/grpc_types.h [hunk 2]
```diff
@@ -213,13 +213,15 @@ typedef struct {
 #define GRPC_ARG_SERVICE_CONFIG "grpc.service_config"
 /** LB policy name. */
 #define GRPC_ARG_LB_POLICY_NAME "grpc.lb_policy_name"
-/** Server name. Not intended for external use. */
```
Comment (ctiller) [inline include/grpc/impl/codegen/grpc_types.h]: Suggest moving to an internal header and out from the public interface if it's not intended for external use.

**Comment by markdroth** [comment]
I'm pretty sure that test failure is unrelated to this PR, since it did not fail when running tests for #9067, which includes all of the changes from this PR.  I've filed #9123 to track this.

**Comment by y-zeng** [comment]
LGTM

**Comment by markdroth** [comment]
Note: @ctiller had some comments on this code in #9067.

**Pull request merged**